### PR TITLE
Improve performance of `svd` and `eigen` of `Diagonal`s

### PR DESCRIPTION
**Pull request by dkarrasch**
This reduces lots of intermediate array allocations, and now even works with (Furlong) units. I'll add a test for that case later.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -672,14 +672,20 @@ end
...
+    d = D.diag
+    s = abs.(d)
+    piv = sortperm(s, rev = true)
+    S = s[piv]
+    Td  = typeof(oneunit(T)/oneunit(T))
+    U = zeros(Td, size(D))
+    Vt = copy(U)
+    for i in 1:length(d)
+        j = piv[i]
+        U[j,i] = d[j] >= zero(T) ? one(Td) : -one(Td)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/diagonal.jl:685]: This does not work for Complex numbers.

Comment (oscardssmith) [reply stdlib/LinearAlgebra/src/diagonal.jl:685]: It should be `U[j, i] = d[j]/norm(d[j])`, right?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/diagonal.jl]: True, thanks.

**Comment by dkarrasch** [comment]
This is a wonderful example why splitting LinearAlgebra and SparseArrays was really useful. See the benchmarks:

```julia
julia> D = Diagonal(randn(1024));

julia> @btime svd($D);
  16.898 ms (10275 allocations: 866.75 KiB) # old behavior with sparse U and Vt

julia> @btime svd($D);
  11.371 ms (2076 allocations: 40.38 MiB) # before this PR: dense output + temporary allocations

julia> @btime svd($D);
  1.506 ms (8 allocations: 16.02 MiB) # this PR: dense output + seemingly no unnecessary allocations
```
While sparse output is good from a memory perspective, writing into an initially empty sparse matrix is slow. I think the same is true for the multiplications listed in https://github.com/JuliaLang/julia/commit/14154fcedc13bd7dffbc76d9c828997653651678#commitcomment-64103782. Honestly, I hadn't seen that coming. 😅  So, I believe the "new" defaults (`similar(::StructuredMatrix, T, size) = zeros()` instead of `spzeros`) are really not bad, plus the old behavior is easy to recover.

Pinging some involved people to make sure they join me in my little github celebration 🎉 : @vtjnash @KristofferC @ViralBShah 😄

**Comment by dkarrasch** [comment]
This also fixes an issue in sorted `eig` of Diagonals, where the `sort!` was acting on `D.diag` directly. I'll prepare a minimal fix suitable for backporting.

**Comment by KristofferC** [comment]
I tried to backport this to 1.6 but got a test error:

```jl
julia> E = eigen(Du)
Eigen{Furlong{0, Float64}, Furlong{1, Float64}, Matrix{Furlong{0, Float64}}, Vector{Furlong{1, Float64}}}
values:
3-element Vector{Furlong{1, Float64}}:
  Furlong{1, Float64}(-0.5153917252601611)
 Furlong{1, Float64}(-0.23809233488008882)
   Furlong{1, Float64}(0.1571713328338422)
vectors:
3×3 Matrix{Furlong{0, Float64}}:
 Furlong{0, Float64}(1.0)  Furlong{0, Float64}(0.0)  Furlong{0, Float64}(0.0)
 Furlong{0, Float64}(0.0)  Furlong{0, Float64}(1.0)  Furlong{0, Float64}(0.0)
 Furlong{0, Float64}(0.0)  Furlong{0, Float64}(0.0)  Furlong{0, Float64}(1.0)

julia> @test Matrix(E) == Du
Error During Test at REPL[55]:1
  Test threw exception
  Expression: Matrix(E) == Du
  AssertionError: p == q
  Stacktrace:
    [1] Furlong{1, Float64}(x::Furlong{0, Float64})
      @ Main.Furlongs ~/julia/test/testhelpers/Furlongs.jl:20
    [2] convert(#unused#::Type{Furlong{1, Float64}}, x::Furlong{0, Float64})
...
```

at https://github.com/JuliaLang/julia/blob/ac1d69302ce8e682bef3108296f5a6c182b62e5f/stdlib/LinearAlgebra/test/diagonal.jl#L427

**Comment by dkarrasch** [comment]
I can reproduce. I think the diagonal multiplication code must have changed between versions, and on v1.6 it cannot multiply `vecs::Matrix{Furlong{0}} * Diagonal(vals::Vector{Furlong{1}}`. It tries to copy `vecs` into a matrix of the result type, which impossible. Long story short, it fails for reasons outside of this commit, so can we still backport but omit the unitful tests?

**Comment by oscardssmith** [comment]
why are we back porting? isn't this just a performance improvement?

**Comment by dkarrasch** [comment]
First of all, it contains a fix for the aliasing of `D.diag` in sorted eigendecompositions.

**Pull request merged**